README.md
# Torproxy 0.9.1 #
A transparent TOR proxying applet for Gnome Desktop. Creates a system-wide proxy that redirects all traffic though TOR.
Based on ParrotSec's anonsurf, with a graphical frontend built as a Gnome Shell Extension.

Currently a functioning prototype. Feel free to download and play with it, but DO NOT rely on this for anonymity.
It's very early alpha! Requires installation of the `libnotify-bin` package to function properly. Made for
Debian-based systems but should work on others with Gnome. May require tweaking of some of (currently) hardcoded
variables depending on your system settings, distribution and tor installation.

**New addition:** launch a custom Torbrowser instance from the menu, configured to run in transparent proxy mode! (requires Torbrowser, obviously).


#### Screenshots ####

![Disabled](https://i.imgsafe.org/8894fd2784.png)
![Enabled](https://i.imgsafe.org/8895640fd3.png)
![Notification](http://i.imgsafe.org/b4fae1a0d4.png)
![Torbrowser](https://i.imgsafe.org/8895ff0f4e.png)


#### Installation: ####

Download and extract. `cd` into the folder and run `sudo ./install.sh`


#### Licence ####

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


#### Acknowledgements ####

Inspired by / heavily plagiarized from the following:

- ParrotSec: https://github.com/ParrotSec/anonsurf
- Und3rf10w: https://github.com/Und3rf10w/kali-anonsurf
- BrainfuckSec: https://github.com/BrainfuckSec/kalitorify
- Subgraph OS: https://github.com/subgraph/gnome-shell-extension-torstatus
- Tor Project: https://wiki.torproject.org/noreply/TheOnionRouter/TransparentProxy


#### TODO ####

- Improve settings panel.
- Improve error handling (if TOR UID is incorrect the error is quietly passed to sys logs, not notify)
- Pull all the logic out of the shell script and into a GSE-style javascript controller
- Refactor.
- Do more to ensure the extension is distribution-agnostic
- Add gnome shell startup checks to see if Torproxy shutdown scripts were run before last system shutdown (do cleanup)
- Make fully translatable.
- Wrap everything up in an installable deb package?
- Find some collaborators. Possibly you, if you're reading this :)

changelog.txt
Version 0.9.1

-Added Gnome Shell Extensions settings panel
-Changed how custom torrc options are loaded (still not great)
-Reverted to using "service tor start" instead of calling tor binary
-Added new launchable Torbrowser configured for transparent proxy mode!

prefs.js
// -*- mode: js2; indent-tabs-mode: nil; js2-basic-offset: 4 -*-

const Gio = imports.gi.Gio;
const GObject = imports.gi.GObject;
const Gtk = imports.gi.Gtk;
const Lang = imports.lang;

const Gettext = imports.gettext.domain('gnome-shell-extension-torproxy');
const _ = Gettext.gettext;

const ExtensionUtils = imports.misc.extensionUtils;
const Me = ExtensionUtils.getCurrentExtension();
const Convenience = Me.imports.convenience;


function init() {
    Convenience.initTranslations();
}

const TorproxyPrefsWidget = new GObject.Class({
    Name: 'Torproxy.Prefs.Widget',
    GTypeName: 'TorproxyPrefsWidget',
    Extends: Gtk.Grid,

    _init: function(params) {
        this.parent(params);

        this.margin = 24;
        this.row_spacing = 6;
        this.orientation = Gtk.Orientation.VERTICAL;
        this._settings = Convenience.getSettings();

        let align = new Gtk.Alignment({ left_padding: 12 });
        this.add(align);

        let heading1 = '<b>' + _("Cache Settings") + '</b>';
        this.add(new Gtk.Label({ label: heading1, use_markup: true, halign: Gtk.Align.START }));

        let clear_caches = new Gtk.CheckButton({ label: _("Clear caches on startup/shutdown (requires Bleachbit)"), margin_top: 6, margin_bottom: 6 });
        this._settings.bind('clear-caches', clear_caches, 'active', Gio.SettingsBindFlags.DEFAULT);
        this.add(clear_caches);

        let heading2 = '<b>' + _("TOR User ID") + '</b>';
        this.add(new Gtk.Label({ label: heading2, use_markup: true, halign: Gtk.Align.START, margin_top: 12 }));

        let tor_uid_label = new Gtk.Label({label: "Should be 'debian-tor' or 'tor', depending on distro", xalign: 0});
        this.add(tor_uid_label);
        let tor_uid = new Gtk.Entry();
        this._settings.bind('tor-uid', tor_uid, 'text', Gio.SettingsBindFlags.DEFAULT);
        this.add(tor_uid);


    }
});

function buildPrefsWidget() {
    let widget = new TorproxyPrefsWidget();
    widget.show_all();

    return widget;
}


install.sh
#!/bin/sh

CURR_USER=`logname`
INSTALL_DIR=/home/$CURR_USER/.local/share/gnome-shell/extensions/torproxy@dot.slash
SCHEMA_DIR=/usr/share/glib-2.0/schemas
SCHEMA_NAME=org.gnome.shell.extensions.torproxy.gschema.xml

#Check permissions

if [ $(id -u) -ne 0 ]; then
    echo 'This install script requires elevated permissions. Try: "sudo ./install.sh" instead.'
    exit 1
fi

#Check for notify binary

if [ ! -e /usr/bin/notify-send ];then
    echo 'Please install notify binary first: "sudo apt-get install libnotify-bin"'
    exit 1
fi

#Install files

rm -r $INSTALL_DIR >&2
mkdir -p $INSTALL_DIR

for file in convenience.js extension.js metadata.json prefs.js reset_prefs.sh install.sh torproxy.sh icons schemas; do
    cp -r ./$file $INSTALL_DIR/
done

#Set up preferences

rm $SCHEMA_DIR/$SCHEMA_NAME >&2
cp ./schemas/$SCHEMA_NAME $SCHEMA_DIR/$SCHEMA_NAME

glib-compile-schemas $SCHEMA_DIR

echo "Installation complete."


torproxy.sh
#!/bin/bash

# Get non-sudo user
CURR_USER=`logname`

# Set up correct user DBUS address for notifications
PID=`pgrep --newest -u $CURR_USER gnome-settings`
DBUS=`grep -z DBUS_SESSION_BUS_ADDRESS /proc/$PID/environ | tr -d '\0' | \
sed -e 's/DBUS_SESSION_BUS_ADDRESS=//'`
export DBUS_SESSION_BUS_ADDRESS=$DBUS

# The UID Tor runs as. Check /usr/share/tor/tor-service-defaults-torrc "User" field for default.
TOR_UID=`gsettings get org.gnome.shell.extensions.torproxy tor-uid | cut -d "'" -f 2`

CLEAR_CACHES=`gsettings get org.gnome.shell.extensions.torproxy clear-caches`

# Tor's VirtualAddrNetworkIPv4
VIRT_ADDR="10.192.0.0/10"

# LAN destinations that shouldn't be routed through Tor
NON_TOR="127.0.0.0/8 10.0.0.0/8 172.16.0.0/12 192.168.0.0/16 127.0.0.0/9 127.128.0.0/10"

# Other IANA reserved blocks (These are not processed by tor and dropped by default)
RESV_IANA="0.0.0.0/8 100.64.0.0/10 169.254.0.0/16 192.0.0.0/24 192.0.2.0/24 192.88.99.0/24 198.18.0.0/15 198.51.100.0/24 203.0.113.0/24 224.0.0.0/3"

# Tor's TransPort
TRANS_PORT="9040"

# DNS redirect port
DNS_PORT="5353"


function notify {
	if [ -e /usr/bin/notify-send ]; then
        local notice=$1
        time=2500
        path="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
        icon=$path"/icons/icon180.png"

        if [ $USER != $CURR_USER ]; then
            su "$CURR_USER" -c "notify-send Torproxy '$notice' -u low -t $time -i $icon"
        else
            notify-send Torproxy "$notice" -u low -t $time -i $icon
        fi
	fi
}

function init {
	#notify "Killing dangerous processes and clearing cache..."

	echo -e "[Torproxy] Killing dangerous applications"
	killall -q chrome dropbox iceweasel skype icedove thunderbird firefox firefox-esr chromium xchat transmission deluge pidgin pidgin.orig

	if $CLEAR_CACHES && hash bleachbit 2>/dev/null; then
	    echo -e "[Torproxy] Clearing some dangerous cache elements"
	    bleachbit -c adobe_reader.cache chromium.cache chromium.current_session chromium.history elinks.history emesene.cache epiphany.cache firefox.url_history flash.cache flash.cookies google_chrome.cache google_chrome.history links2.history opera.cache opera.search_history opera.url_history &> /dev/null
    fi
}

function start {
	check_root

	echo -e "[Torproxy] Starting Torproxy..."
    notify "Starting Torproxy..."

	disable_ufw

	# Kill IPv6 services
	echo -e "[Torproxy] Stopping IPv6 services"
	sed -i '/^.*\#Torproxy$/d' /etc/sysctl.conf
	echo "net.ipv6.conf.all.disable_ipv6=1 #Torproxy" >> /etc/sysctl.conf
	echo "net.ipv6.conf.default.disable_ipv6=1 #Torproxy" >> /etc/sysctl.conf
	sysctl -p > /dev/null

    service network-manager force-reload > /dev/null 2>&1
    service resolvconf stop 2>/dev/null || true
    service nscd stop 2>/dev/null || true
    service dnsmasq stop 2>/dev/null || true
    sleep 1
    killall dnsmasq nscd resolvconf 2>/dev/null || true
    sleep 1
    service resolvconf start 2>/dev/null || true

    # Append config options and some socks safety hacks to torrc file
    sed -i '/^.*\#Torproxy$/d' /etc/tor/torrc
    echo "VirtualAddrNetworkIPv4 10.192.0.0/10 #Torproxy" >> /etc/tor/torrc
    echo "AutomapHostsOnResolve 1 #Torproxy" >> /etc/tor/torrc
    echo "TransPort 9040 #Torproxy" >> /etc/tor/torrc
    echo "SocksPort 9050 #Torproxy" >> /etc/tor/torrc
    #echo "ControlPort 9151 #Torproxy" >> /etc/tor/torrc
    echo "DNSPort 5353 #Torproxy" >> /etc/tor/torrc
    echo "TestSocks 1 #Torproxy" >> /etc/tor/torrc
    echo "SafeSocks 1 #Torproxy" >> /etc/tor/torrc
    echo "WarnPlaintextPorts 23,109,110,143,80 #Torproxy" >> /etc/tor/torrc

    #/usr/bin/tor --defaults-torrc /usr/share/tor/tor-service-defaults-torrc -f /home/$CURR_USER/.local/share/gnome-shell/extensions/torproxy@dot.slash/torrc --RunAsDaemon 1

    # Start TOR service
    service tor start
    #sleep 1

    # Save IP table rules
	if [ ! -f /etc/network/iptables.rules ]; then
		iptables-save > /etc/network/iptables.rules
		echo -e "[Torproxy] Saved iptables rules"
	fi

	# Flush IP tables
	iptables -F
	iptables -t nat -F

	# Set DNS resolver to use TOR
	cp -vf /etc/resolv.conf /etc/resolv.conf.backup
    echo -e 'nameserver 127.0.0.1' > /etc/resolv.conf

    # Don't nat the TOR process itself
    iptables -t nat -A OUTPUT -m owner --uid-owner $TOR_UID -j RETURN

	# Nat DNS requests to TOR
	iptables -t nat -A OUTPUT -p udp --dport 53 -j REDIRECT --to-ports $DNS_PORT
	iptables -t nat -A OUTPUT -p tcp --dport 53 -j REDIRECT --to-ports $DNS_PORT
	iptables -t nat -A OUTPUT -p udp -m owner --uid-owner $TOR_UID -m udp --dport 53 -j REDIRECT --to-ports $DNS_PORT

	# Nat .onion addresses
	iptables -t nat -A OUTPUT -p tcp -d $VIRT_ADDR -j REDIRECT --to-ports $TRANS_PORT
	iptables -t nat -A OUTPUT -p udp -d $VIRT_ADDR -j REDIRECT --to-ports $TRANS_PORT
	
	# Exclude local network addresses
	for LAN in $NON_TOR; do
		iptables -t nat -A OUTPUT -d $LAN -j RETURN
		iptables -A OUTPUT -d $LAN -j ACCEPT
	done

	# Exclude IANA reserved blocks
	for IANA in $RESV_IANA; do
        iptables -t nat -A OUTPUT -d $IANA -j RETURN
    done
	
	# Redirect all other output through TOR
	iptables -t nat -A OUTPUT -p tcp --syn -j REDIRECT --to-ports $TRANS_PORT
	iptables -t nat -A OUTPUT -p udp -j REDIRECT --to-ports $TRANS_PORT
	iptables -t nat -A OUTPUT -p icmp -j REDIRECT --to-ports $TRANS_PORT
	
	# Accept already established connections
	iptables -A OUTPUT -m state --state ESTABLISHED,RELATED -j ACCEPT

	# Allow only TOR output
	iptables -A OUTPUT -m owner --uid-owner $TOR_UID -j ACCEPT
	iptables -A OUTPUT -j REJECT

    # Filter input
    iptables -A INPUT -m state --state ESTABLISHED -j ACCEPT
    iptables -A INPUT -i lo -j ACCEPT
	iptables -A INPUT -j DROP

	# Filter output
	iptables -A OUTPUT -m state --state INVALID -j DROP
	iptables -A OUTPUT -d 127.0.0.1/32 -o lo -j ACCEPT

	# Set default policies to DROP
    iptables -P INPUT DROP
    iptables -P FORWARD DROP
    iptables -P OUTPUT DROP

	echo -e "[Torproxy] All traffic is now redirected through TOR"
	notify "Torproxy is up and running."
}


function stop {
	check_root

    notify "Shutting down Torproxy..."
	echo -e "[Torproxy] Stopping Torproxy..."

    # Flush iptables
	iptables -F
	iptables -t nat -F
	echo -e "[Torproxy] Flushed iptables rules"

	# Restore iptable rules
	if [ -f /etc/network/iptables.rules ]; then
		iptables-restore < /etc/network/iptables.rules
		rm /etc/network/iptables.rules
		echo -e "[Torproxy] Iptables rules restored"
	fi
	
	# Restore DNS settings
	if [ -e /etc/resolv.conf.backup ]; then
		rm /etc/resolv.conf
		cp /etc/resolv.conf.backup /etc/resolv.conf
	fi

    # Stop TOR service
	service tor stop >&2
	sleep 1
	#killall tor >&2

	# Remove custom config options from torrc file
	sed -i '/^.*\#Torproxy$/d' /etc/tor/torrc

	# Re-enable IPv6 services
	echo -e "[Torproxy] Reenabling IPv6 services"
	sed -i '/^.*\#Torproxy$/d' /etc/sysctl.conf
	sysctl -p

    # Reload services
	service network-manager force-reload > /dev/null 2>&1
	service nscd start > /dev/null 2>&1
	service dnsmasq start > /dev/null 2>&1
	
	echo -e "[Torproxy] Anonymous mode stopped"
    notify "Torproxy is now disabled."

    enable_ufw
}

function change {
	service tor reload
	sleep 1
	echo -e "[Torproxy] TOR daemon reloaded and forced to change nodes"

	notify "TOR service reloaded with a fresh circuit"
}

function status {
    status=`/etc/init.d/tor status | grep Active: | sed -e 's/^.*\Active://'`
	notify "Status: $status"
}

function check_root {
    if [ $(id -u) -ne 0 ]; then
		notify "Torproxy must be run as root."
		echo -e "[Torproxy] Torproxy must be run as root" >&2
		exit 1
	fi
}

function disable_ufw () {
	if hash ufw 2>/dev/null; then
    	if ufw status | grep -q active$; then
        	echo -e "[Torproxy] Disabling ufw firewall..."
        	ufw disable
        	sleep 6
    	fi
    fi
}

function enable_ufw () {
	if hash ufw 2>/dev/null; then
    	if ufw status | grep -q inactive$; then
        	echo -e "[Torproxy] Re-enabling ufw firewall..."
        	ufw enable
        fi
    fi
}

case "$1" in
	start)
		init
		start
	;;
	stop)
		init
		stop
	;;
	change)
		change
	;;
	status)
		status
	;;
	restart)
		stop
		sleep 2
		start
	;;
    *)
        exit 1
    ;;
esac

exit 0


convenience.js
/* -*- mode: js; js-basic-offset: 4; indent-tabs-mode: nil -*- */
/*
  Copyright (c) 2011-2012, Giovanni Campagna <scampa.giovanni@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:
    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
    * Neither the name of the GNOME nor the
      names of its contributors may be used to endorse or promote products
      derived from this software without specific prior written permission.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

const Gettext = imports.gettext;
const Gio = imports.gi.Gio;

const Config = imports.misc.config;
const ExtensionUtils = imports.misc.extensionUtils;

/**
 * initTranslations:
 * @domain: (optional): the gettext domain to use
 *
 * Initialize Gettext to load translations from extensionsdir/locale.
 * If @domain is not provided, it will be taken from metadata['gettext-domain']
 */
function initTranslations(domain) {
    let extension = ExtensionUtils.getCurrentExtension();

    domain = domain || extension.metadata['gettext-domain'];

    // check if this extension was built with "make zip-file", and thus
    // has the locale files in a subfolder
    // otherwise assume that extension has been installed in the
    // same prefix as gnome-shell
    let localeDir = extension.dir.get_child('locale');
    if (localeDir.query_exists(null))
        Gettext.bindtextdomain(domain, localeDir.get_path());
    else
        Gettext.bindtextdomain(domain, Config.LOCALEDIR);
}

/**
 * getSettings:
 * @schema: (optional): the GSettings schema id
 *
 * Builds and return a GSettings schema for @schema, using schema files
 * in extensionsdir/schemas. If @schema is not provided, it is taken from
 * metadata['settings-schema'].
 */
function getSettings(schema) {
    let extension = ExtensionUtils.getCurrentExtension();

    schema = schema || extension.metadata['settings-schema'];

    const GioSSS = Gio.SettingsSchemaSource;

//    // check if this extension was built with "make zip-file", and thus
//    // has the schema files in a subfolder
//    // otherwise assume that extension has been installed in the
//    // same prefix as gnome-shell (and therefore schemas are available
//    // in the standard folders)
//    let schemaDir = extension.dir.get_child('schemas');
//    let schemaSource;
//    if (schemaDir.query_exists(null))
//        schemaSource = GioSSS.new_from_directory(schemaDir.get_path(),
//                                                 GioSSS.get_default(),
//                                                 false);
//    else
//        schemaSource = GioSSS.get_default();

    let schemaSource = GioSSS.get_default();

    let schemaObj = schemaSource.lookup(schema, true);
    if (!schemaObj)
        throw new Error('Schema ' + schema + ' could not be found for extension '
                        + extension.metadata.uuid + '. Please check your installation.');

    return new Gio.Settings({ settings_schema: schemaObj });
}
								  


extension.js
const St = imports.gi.St;
const Main = imports.ui.main;

const GLib = imports.gi.GLib;
const Gio = imports.gi.Gio;
const Gdk = imports.gi.Gdk;

const Shell = imports.gi.Shell;
const Util = imports.misc.util;
const Lang = imports.lang;
const PopupMenu = imports.ui.popupMenu;
const PanelMenu = imports.ui.panelMenu;

let TorproxyInstance, path, theme;

function init(extensionMeta) {
    // Set path for custom icon
    path = extensionMeta.path;
    theme = imports.gi.Gtk.IconTheme.get_default();
    theme.append_search_path(path + "/icons");
}

function enable() {
    TorproxyInstance = new TorproxyMenu(path);
    Main.panel.addToStatusArea('Torproxy', TorproxyInstance);
}

function disable() {
    TorproxyInstance.destroy();
    TorproxyInstance = null;
}

const TorproxyMenu = new Lang.Class({
    Name: 'TorProxy',
    Extends: PanelMenu.Button,

    _init: function(path) {
        this.parent(null, 'TorProxy');

        this.user = GLib.get_user_name();
        this.filepath = path;
        this.active = false;

        this.icon_inactive = 120;
        this.icon_active = 235;

        this.addPanelButton();
        this.createMenu();
    },

    addPanelButton: function() {
        // Create panel button
        this.button = new St.BoxLayout({ style_class: "main-button" });
        this.icon = new St.Icon({icon_name: 'tor-simple-symbolic', style_class: 'system-status-icon'});
        this.icon.opacity = this.icon_inactive;
        this.button.add_actor(this.icon);
        this.actor.add_actor(this.button);

        // Change icon opacity on hover
        this.actor.connect('enter_event', Lang.bind(this, function() {
            this.icon.opacity += 20;
        }));
        this.actor.connect('leave_event', Lang.bind(this, function() {
            this.icon.opacity -= 20;
        }));
    },

    createMenu: function() {
        // Add menu items
        this.torproxySwitch = new PopupMenu.PopupSwitchMenuItem("Torproxy");
        this.torproxySwitch.connect('toggled', Lang.bind(this, this.toggleSwitch));
        this.menu.addMenuItem(this.torproxySwitch);

        this.menu.addMenuItem(new PopupMenu.PopupSeparatorMenuItem);

        this.menuChange = new PopupMenu.PopupMenuItem("New TOR circuit");
	    this.menuChange.actor.reactive = false;
	    this.menuChange.connect('activate', Lang.bind(this, this._tpChange));
	    this.menu.addMenuItem(this.menuChange);

	    this.menuRestart = new PopupMenu.PopupMenuItem("Restart Torproxy");
	    this.menuRestart.actor.reactive = false;
	    this.menuRestart.connect('activate', Lang.bind(this, this._tpRestart));
	    this.menu.addMenuItem(this.menuRestart);

	    this.menuCheck = new PopupMenu.PopupMenuItem("Check TOR status");
	    this.menuCheck.actor.reactive = false;
	    this.menuCheck.connect('activate', Lang.bind(this, this._tpCheck));
	    this.menu.addMenuItem(this.menuCheck);

	    this.menuBrowser = new PopupMenu.PopupMenuItem("Launch Torbrowser");
	    this.menuBrowser.actor.reactive = false;
	    this.menuBrowser.connect('activate', Lang.bind(this, this._tpBrowser));
	    this.menu.addMenuItem(this.menuBrowser);
    },

    toggleSwitch: function() {
        if (this.active) {
            this._tpStop();
        } else {
            this._tpStart();
        }
    },

    setState: function(newState) {
        //TODO: store the state in a preference, check on startup and clean up / restore system settings as needed
        switch (newState) {
            case 'on':
                this.icon.opacity = this.icon_active;
                this.active = true;
                this.torproxySwitch.setToggleState(true);
                this.menuChange.actor.reactive = true;
                this.menuRestart.actor.reactive = true;
                this.menuCheck.actor.reactive = true;
                this.menuBrowser.actor.reactive = true;
                break;
            case 'off':
                this.icon.opacity = this.icon_inactive;
                this.active = false;
                this.torproxySwitch.setToggleState(false);
                this.menuChange.actor.reactive = false;
                this.menuRestart.actor.reactive = false;
                this.menuCheck.actor.reactive = false;
                this.menuBrowser.actor.reactive = false;
                break;
        }
    },

    _tpStart: function() {
        this._spawn(['pkexec', this.filepath+'/torproxy.sh', 'start'], 'on');
    },

    _tpStop: function() {
        //TODO: double-check to see if Torbrowser prefs backup needs restoring?
        this._spawn(['pkexec', this.filepath+'/torproxy.sh', 'stop'], 'off');
    },

    _tpChange: function() {
        this._spawn(['pkexec', this.filepath+'/torproxy.sh', 'change']);
    },

    _tpRestart: function() {
        this._spawn(['pkexec', this.filepath+'/torproxy.sh', 'restart']);
    },

    _tpCheck: function() {
        this._spawn([this.filepath+'/torproxy.sh', 'status']);
    },

    _tpBrowser: function() {
        //TODO: Check that torbrowser is actually installed... disable menu option if not.
        //TODO: Check browser isn't open already, deal with it if so...?
        this._notify('Starting Torbrowser in transparent proxy mode...');

        // Back up Torbrowser preferences before launch (very important!)
        if(this._tpBackupPrefs()) {
            //TODO: set up control port for multiple TOR circuits

            // Basic Torbrowser command
            let cmd = [
                '/home/'+this.user+'/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/firefox',
                '--class Tor Browser',
                '-profile TorBrowser/Data/Browser/profile.default'
            ];

            // Add environment variables (skip Vidalia on launch, etc)
            let base_env = GLib.get_environ();
            let cust_env = [
                'TOR_SKIP_LAUNCH=1',
                'TOR_TRANSPROXY=1',
                'TOR_SOCKS_HOST=10.192.0.1',
                'TOR_SOCKS_PORT=9050',
                'TOR_SKIP_CONTROLPORTTEST=1',
                //'TOR_CONTROL_PORT=9151',
                //'TOR_CONTROL_COOKIE_AUTH_FILE=/var/run/tor/control.authcookie',
                'TOR_NO_DISPLAY_NETWORK_SETTINGS=1'
            ];

            // Spawn the process
            let [result, child_pid] = GLib.spawn_async(
                null, //str working directory
                cmd, //arr args
                base_env.concat(cust_env), //arr env
                GLib.SpawnFlags.DO_NOT_REAP_CHILD, //flags spawn flags
                null //func child setup function
            );

            // Disable "Launch Torbrowser" button
            this.menuBrowser.actor.reactive = false;

            // Attatch callback (called when process ends)
            GLib.child_watch_add(GLib.PRIORITY_DEFAULT, child_pid, Lang.bind(this, function(pid, exitStatus) {
                // Resore preferences backup when Torbrowser exit code received
                this._notify("Shutting down transparent proxy Torbrowser...");
                this._tpRestorePrefs();
                GLib.spawn_close_pid(pid);
                // Re-enable button
                this.menuBrowser.actor.reactive = true;
            }));

        } else {
            // Failed to back up pref.js (don't open the browser!)
            this._notify("Failed to back up Torbrowser system preferences");
        }
    },

    _tpBackupPrefs: function() {
        let prefs_path = '/home/'+this.user+'/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/profile.default/prefs.js';
        let prefs_file = Gio.File.new_for_path(prefs_path);
        let prefs_backup = Gio.File.new_for_path(prefs_path+'.backup');

        if(prefs_backup.query_exists(null)) {
            //A backup already exists... don't overwrite it.
            return true;
        }

        if(prefs_file.query_exists(null)) {
            //Save backup.

            //This messy prefs.js backup hack is important as Torbrowser currently alters it's settings when using
            //the 'TOR_SKIP_LAUNCH=1' and 'TOR_TRANSPROXY=1' options, but they don't get changed back. This can leave
            //the Torbrowser broken when launching it normally with Vidalia when not using the transparent proxy.
            //https://trac.torproject.org/projects/tor/ticket/17615

            if(prefs_file.copy(prefs_backup, Gio.FileCopyFlags.OVERWRITE, null, null, null)) {
                this._notify("Backed up Torbrowser system preferences");
                return true;
            } else {
                //this._notify("Failed to back up Torbrowser preferences"); //Handed elsewhere
            }
        }

        return false;
    },

    _tpRestorePrefs: function() {
        let prefs_path = '/home/'+this.user+'/.local/share/torbrowser/tbb/x86_64/tor-browser_en-US/Browser/TorBrowser/Data/Browser/profile.default/prefs.js';
        let prefs_file = Gio.File.new_for_path(prefs_path);
        let prefs_backup = Gio.File.new_for_path(prefs_path+'.backup');

        if(prefs_backup.query_exists(null)) {
            //Restore prefs from backup
            if(prefs_backup.copy(prefs_file, Gio.FileCopyFlags.OVERWRITE, null, null, null)) {
                //Delete backup
                prefs_backup.delete_async(GLib.PRIORITY_DEFAULT, null, null);
                this._notify("Restored Torbrowser system preferences");
                return true;
            }
        } else {
            // Backup not found
            this._notify("Failed to restore Torbrowser system preferences: backup file not found");
        }

        return false;
    },

    _spawn: function(argv, newState) {
        try {
            // Spawn the process
            let [result, child_pid] = GLib.spawn_async(
                null, //str working directory
                argv, //arr arguments
                null, //arr environment variables
                GLib.SpawnFlags.SEARCH_PATH | GLib.SpawnFlags.DO_NOT_REAP_CHILD, //flags spawn flags
                null //func child setup function
            );

            // Set callback for process completion
            GLib.child_watch_add(GLib.PRIORITY_DEFAULT, child_pid,
                Lang.bind(this, function(pid, exitStatus, requestObj) {

                // If changing Torproxy active state
                if(newState != null && exitStatus == "0") {
                    // Change Torproxy state if successful
                    this.setState(newState);
                } else if (newState != null && exitStatus != "0") {
                    //Reset switch class if process failed
                    switch (newState) {
                        case 'on':
                            this.torproxySwitch.setToggleState(false);
                            break;
                        case 'off':
                            this.torproxySwitch.setToggleState(true);
                            break;
                    }
                }

                // Clean up when finished.
                GLib.spawn_close_pid(pid);
            }));

        } catch (err) {
            // Handle errors
            if (err.matches(GLib.SpawnError, GLib.SpawnError.NOENT)) {
                throw new GLib.SpawnError({ code: GLib.SpawnError.NOENT, message: _("Command not found") });
            } else if (err instanceof GLib.Error) {
                message = err.message.replace(/.*\((.+)\)/, '$1');
                throw new (err.constructor)({ code: err.code, message: message });
            } else {
                throw err;
            }
        }
    },

    _notify: function(notifyMessage) {
        global.log("[Torproxy] "+notifyMessage);
        GLib.spawn_command_line_async("notify-send Torproxy '"+notifyMessage+"' -u low -t 2500 -i "+path+"/icons/icon180.png");
    },

    destroy: function() {
        this.parent();
    },

});
